### ISSUE TITLE
Implement unrolled linked lists

### ISSUE THREAD

**Issue by PallHaraldsson** [JuliaCollections/DataStructures.jl#339]
I proposed at https://github.com/adolgert/Lists.jl/issues/1

I.e. https://en.wikipedia.org/wiki/Unrolled_linked_list

See e.g.:
http://www.sanfoundry.com/java-program-implement-unrolled-linked-list/

And got this response:

"That seems like a great idea."

I couldn't find the issue here; I thought I had filed here, and see now the other project seems dead.

**Comment by oxinabox** [comment]
Isn't this what our Deque is?
(honest question)

https://github.com/JuliaCollections/DataStructures.jl/blob/master/src/deque.jl

**Comment by kmsquire** [comment]
Yes, that implementation is pretty much equivalent.

**Comment by PallHaraldsson** [comment]
"Isn't this what our Deque is?" Ok, great then, I didn't realize as a deque can be implemented another way, e.g. non-unrolled list. If you can really do all the same things as with a list, e.g. add or drop from any position, then this issue can be closed; but maybe document then better that it can work as/is an unrolled linked-list.

**Comment by PallHaraldsson** [comment]
There's also at least on generalization: https://en.wikipedia.org/wiki/T-tree

I see 2-3 tree in the code; is there no 3+ tree in the code, and isn't it better to have? And deque only blocked/unrolled list? I so on WP:

the B-tree and T-tree, data structures that are similar to unrolled linked lists in the sense that each of them could be viewed as an "unrolled binary tree"

**Comment by kmsquire** [comment]
> If you can really do all the same things as with a list, e.g. add or drop from any position,

The underlying `DequeBlock` data structure is [here](https://github.com/JuliaCollections/DataStructures.jl/blob/master/src/deque.jl#L8-L23), and the corresponding `Deque` implementation is below that.

`DequeBlock` and `Deque` could, in theory, be generalized to add or drop from any position, but there is no functionality to do so right now.  Some comments:

1.  Arguments were made by @lindahua, the original author of this code, that, for `Deque` (and `Stack` and `Queue`), there is expected functionality (push/pop/enqueue/dequeue) for each data structure, and that it's better to keep the implementations simple, in order to keep them performant.  One could argue differently of course.
   (This is a paraphrase from some old conversations here that I don't want to take the time to find right now, so @lindahua, please speak up if I'm misrepresenting you!)

2. If the functionality were to change, it would be good maintain a consistent interface among related data structures (e.g., `Deque`, `Stack`, and `Queue` here).

**Comment by kmsquire** [comment]
> I see 2-3 tree in the code; is there no "3+" tree in the code, and isn't it better to have?

If it (or anything else you're proposing) is useful to you, please implement and submit a PR!  There are a lot of things that could be improved about the code here, but as with most open source projects, it depends on someone interested and motivated enough to implement them in their spare time.

**Comment by StephenVavasis** [comment]
There is a work-in-progress pull request to implement cache-oblivious B-trees here: https://github.com/JuliaCollections/DataStructures.jl/pull/241 This would be a drop-in replacement for the 2-3 trees used in the sorted containers.  The idea is to get better performance by minimizing cache misses.

**Comment by kmsquire** [comment]
@StephenVavasis, are you working with @alooow, the author of that code?  She hasn't made any updates since July.

**Comment by StephenVavasis** [comment]
No, I'm not working with her.  Occasionally I made some suggestions about her code via github, but I have no contact with her outside of github.
